src/array.rs

```diff
@@ -21,7 +20,8 @@ impl TypeArray for ATerm {}
 /// may find it lacking functionality.
 #[derive(Eq, PartialEq, Ord, PartialOrd, Clone, Copy, Hash, Debug)]
 pub struct TArr<V, A> {
-    _marker: PhantomData<(V, A)>,
+    first: V,
+    rest: A,
 }
```

Comment:
The structure of `TArr` is changed so that `V` and `A` exist in the value domain. This allows us to compute on values of `V` and `A`, and remove the uses of `unsafe` that were needed to construct values out of thin air.